### PR TITLE
[SDPA-2449] change string long to text long

### PR DESCRIPTION
**Pull request by vincent-gao**
https://digital-engagement.atlassian.net/browse/SDPA-2449

### CODE REVIEW

**Comment by GROwen** [comment]
@vincent-gao I've updated the base branch to `develop` (I'll update the default branch to `develop` while I'm at it). Tests are failing because of an update that's pending in #22 - which means this PR can be tested.

File: tide_media.install [hunk 1]
```diff
@@ -150,3 +150,81 @@ function tide_media_update_8002() {
...
+      continue;
+    }
+
+    $new_fields = [];
+    foreach ($field_storage->getBundles() as $bundle => $label) {
+      $field = FieldConfig::loadByName($entity_type, $bundle, $field_name);
+      $new_field = $field->toArray();
+      $new_field['field_type'] = 'text_long';
+      $new_fields[] = $new_field;
+      $field->delete();
```
Comment (sonnykt) [inline tide_media.install]: @vincent-gao as the field tables are truncated, the field itself has no data. You should be able to change its field type directly without deleting and re-creating the field.

File: tide_media.install [hunk 2]
```diff
@@ -150,3 +150,81 @@ function tide_media_update_8002() {
...
+      $new_field->save();
+    }
+
+    //step 3 data migration
+    foreach ($tmp_table_names as $tmp => $table_name) {
+      $database->schema()->addField($tmp,'field_media_transcript_format',$spec);
+      $database->update($tmp)->fields(['field_media_transcript_format'=>'rich_text'])->execute();
+      $database->insert($table_name)->from($database->select($tmp, 't')
+        ->fields('t'))->execute();
+      $query = 'UPDATE `'."$table_name" .'` SET `field_media_transcript_value` = REPLACE(`field_media_transcript_value`, \'\r\n\', \'<br />\')';;
```
Comment (sonnykt) [inline tide_media.install]: @vincent-gao line-breaks are not always `\r\n`. `\n` and `\r` are also line-breaks. Please use `php nl2br()` instead.

File: tide_media.install (hunk 1 above)
Comment (vincent-gao) [reply tide_media.install]: Hi @sonnykt ,  if the UUID didn't change, we don't need to keep fixating on `change it field type`.

File: tide_media.install (hunk 2 above)
Comment (vincent-gao) [reply tide_media.install]: Hi @sonnykt , I prefer use mysql query directly. based on the reason you metioned before, I dont want to load all values into php.
I will update the query.

File: tide_media.install (hunk 1 above)
Comment (sonnykt) [reply tide_media.install]: @vincent-gao I'm hesitate to delete fields unless it's necessary. Deleting a field may cause its dependants being updated/deleted. For example, deleting a field will **disable** all views using that view in Fields, Filters, Relationships, etc. As Tide Media is an opensource project, we can't be sure if there is any dependants of the field in other independent sites.

image

Comment (sonnykt) [reply tide_media.install]: Also, if you delete the field, it's pointless to truncate the field tables as they will be deleted anyway.

File: tide_media.install [hunk 3]
```diff
@@ -156,7 +156,12 @@ function tide_media_update_8002() {
  */
 function tide_media_update_8003() {
   $database = Drupal::database();
-
+  $line_breaks = [
+    '\'\r\n\'',
+    '\'\n\r\'',
```
Comment (sonnykt) [inline tide_media.install]: @vincent-gao `\n\r` is not a linebreak. Valid linebreaks are `\r\n` (CRLF - Windows), `\r` (CR - MacOS 9), and `\n` (LF - *nix).

File: tide_media.install [hunk 4]
```diff
@@ -150,3 +150,79 @@ function tide_media_update_8002() {
     }
   }
 }
+
+/**
+ * Video transcript field change from text field to WYSIWYG
```
Comment (sonnykt) [inline tide_media.install]: Comment must end with a period `.`.

File: tide_media.install [hunk 5]
```diff
@@ -150,3 +150,79 @@ function tide_media_update_8002() {
...
+ */
+function tide_media_update_8003() {
+  $database = Drupal::database();
+  $line_breaks = [
+    '\'\r\n\'',
+    '\'\n\r\'',
+    '\'\n\'',
+    '\'\r\'',
+  ];
+  //step1 create two temporary tables for holding data
```
Comment (sonnykt) [inline tide_media.install]: Inline comments must start with a single space, the first letter must be capitalised, and the sentence must end with a period `.`.

File: tide_media.install (hunk 3 above)
Comment (vincent-gao) [reply tide_media.install]: Thanks.

File: tide_media.install (hunk 4 above)
Comment (vincent-gao) [reply tide_media.install]: thanks. fixed.

File: tide_media.install [hunk 6]
```diff
@@ -150,3 +150,78 @@ function tide_media_update_8002() {
...
+  // Step 3 data migration.
+  foreach ($tmp_table_names as $tmp => $table_name) {
+    $database->schema()->addField($tmp, 'field_media_transcript_format', $spec);
+    $database->update($tmp)
+      ->fields(['field_media_transcript_format' => 'rich_text'])
+      ->execute();
+    $database->insert($table_name)->from($database->select($tmp, 't')
+      ->fields('t'))->execute();
+    foreach ($line_breaks as $line_break) {
+      $query = 'UPDATE `' . "$table_name" . '` SET `field_media_transcript_value` = REPLACE(`field_media_transcript_value`,  ' . $line_break . ', \'<br />\')';;
```
Comment (sonnykt) [inline tide_media.install]: Please check the coding standard, the above line ends with 2 semi-colons `;;`.

IMO dynamic queries with expression are more preferable for UPDATE statements https://www.drupal.org/docs/8/api/database-api/dynamic-queries/introduction-to-dynamic-queries

```suggestion
      $query = $database->update($table_name)->expression('field_media_transcript_value', 'REPLACE(field_media_transcript_value, :linebreak)', [':linebreak' => $line_break])->execute();
```

Comment (vincent-gao) [reply tide_media.install]: the code has been updated.  @sonnykt

File: tide_media.install [hunk 7]
```diff
@@ -243,3 +243,81 @@ function tide_media_update_8005() {
     }
   }
 }
+
+/**
+ * Video transcript field change from text field to WYSIWYG.
+ */
+function tide_media_update_8003() {
```
Comment (sonnykt) [inline tide_media.install]: @vincent-gao pls update the hook number.

**Comment by anthony-malkoun** [comment]
@vincent-gao what's happening with this PR? Please check with @sonnykt that he's happy with your updates as it looks like you've addressed everything he asked about.

**Comment by vincent-gao** [comment]
> @vincent-gao what's happening with this PR? Please check with @sonnykt that he's happy with your updates as it looks like you've addressed everything he asked about.

ok.thanks @anthony-malkoun

File: tide_media.install [hunk 8]
```diff
@@ -243,3 +244,81 @@ function tide_media_update_8005() {
...
+  // Step 3 data migration.
+  foreach ($tmp_table_names as $tmp => $table_name) {
+    $database->schema()->addField($tmp, 'field_media_transcript_format', $spec);
+    $database->update($tmp)
+      ->fields(['field_media_transcript_format' => 'rich_text'])
+      ->execute();
+    $database->insert($table_name)->from($database->select($tmp, 't')
+      ->fields('t'))->execute();
+    foreach ($line_breaks as $line_break) {
+      $database->update('media__field_media_transcript')->expression('field_media_transcript_value', "REPLACE(field_media_transcript_value,:line_break,:br)", [
```
Comment (sonnykt) [inline tide_media.install]: @vincent-gao do we also need to update the `media_revision__field_media_transcript` table?

Comment (vincent-gao) [reply tide_media.install]: @sonnykt yes. typo issue. thanks Sonny.

**Pull request merged**